lib/ddtrace/transport/traces.rb

```diff
@@ -8,25 +9,173 @@ module Traces
       class Parcel
         include Transport::Parcel
 
-        def count
-          data.length
+        attr_reader :trace_count
+
+        def initialize(data, trace_count)
+          super(data)
+          @trace_count = trace_count
         end
 
-        def encode_with(encoder)
-          encoder.encode_traces(data)
+        def count
+          data.length
         end
       end
 
       # Traces request
       class Request < Transport::Request
-        def initialize(traces)
-          super(Parcel.new(traces))
+        def initialize(data, trace_count)
+          super(Parcel.new(data, trace_count))
         end
       end
 
       # Traces response
       module Response
-        attr_reader :service_rates
+        attr_reader :service_rates, :trace_count
+      end
+
+      # Traces chunker
+      class Chunker
+        # Trace agent limit payload size of 10 MiB (since agent v5.11.0):
+        # https://github.com/DataDog/datadog-agent/blob/6.14.1/pkg/trace/api/api.go#L46
+        #
+        # We set the value to a conservative 5 MiB, in case network speed is slow.
+        DEFAULT_MAX_PAYLOAD_SIZE = 5 * 1024 * 1024
+
+        attr_reader :encoder, :max_size
+
+        #
+        # Single traces larger than +max_size+ will be discarded.
+        #
+        # @param encoder [Datadog::Encoding::Encoder]
+        # @param max_size [String] maximum acceptable payload size
+        def initialize(encoder, max_size: DEFAULT_MAX_PAYLOAD_SIZE)
+          @encoder = encoder
+          @max_size = max_size
+        end
+
+        # Encodes a list of traces in chunks.
+        # Before serializing, all traces are normalized. Trace nesting is not changed.
+        #
+        # @param traces [Array<Trace>] list of traces
+        # @return [Enumerator[Array[Bytes,Integer]]] list of encoded chunks: each containing a byte array and
+        #   number of traces
+        def encode_in_chunks(traces)
+          encoded_traces = traces.map { |t| encode_one(t) }.compact
```

Comment:
Not something that must be resolved in this PR but something to think about: should encoding be done as chunks are generated and submitted?

My thinking is if we spend the CPU time on encoding a batch of traces, we should definitely attempt sending it, as we've already paid the cost for it. I think there's a few of ways to go about this:

 - On each chunk, do the encoding only for that chunk and `yield` back to do a request (similar to what you had before.)
 - Move the use of `Chunker` entirely outside the transport; on each chunk, `yield` back a chunk of traces, then `send_traces` with that chunk to the transport. This might be a bigger change because on each chunk, you'd presumably have to encode first before `send_traces`, so we'd have to change encoding based on what encoder is active on the transport at the time. (That would involve extracting a lot of encoding behavior out of the transport.)

Ultimately I think it'd make more sense to end up somewhere like the second suggestion, but the first might be an okay stopgap. Neither though have to be done in this PR... just a consideration for the future.

Reply:
The important piece is that we need to encode each trace individually to know how it'll fit into the allowed max buffer size, and then join them together after based on the encoding method used.

I think the way this reads to me is:

- Encode each trace individually
- Group the traces into chunks based on the max allowed payload size
- Send all chunks to the agent
  - Regardless of how many traces are in the chunks, e.g. sending 1 trace yields 1 chunk and 1 request to the agent

Which if ☝️ is right then it is correct behavior.

We could probably optimize for, encode and generate chunks, and send a request before encoding any more traces to do minimal encoding overhead between requests, but it shouldn't matter too much. Just whether we encode all up front then send, or encode in chunks as we make requests, should be roughly the same CPU overhead either way.

Reply:
I can easily change it to a stream model:
1. Encode as many traces as it fits in a chunk
2. Send chunk
3. Repeat 1 with next chunk

I didn't add it to the PR changes for simplicity (as we already kept all traces serialized in memory before, so the PR is not increasing overahead), but I can add streaming now.

Reply:
Done by using a lazy iterator, which will process one request at a time, in a stream fashion, without ever creating intermediate lists.

The end result is still a list with all Response objects.